### ISSUE TITLE
Using Dokku-alt with Drupal

### ISSUE THREAD

**Issue by that0n3guy** [dokku-alt/dokku-alt#142]
This issue is created so we don't clutter #139.

 My drupal sites aren't yet migrated to dokku, I'm still in the process.   That said, I would be using a Dockerfile and do it something like so: 
- stick the whole drupal site in git
- add dockerfile, use: https://registry.hub.docker.com/u/that0n3guy/baseimage-nginx-phpfpm/ as a base
- add a .docker folder to the root of drupal
- my dockerfile would look something like:

```
FROM that0n3guy/baseimage-nginx-phpfpm

ADD . /app/
RUN rm /app/Dockerfile

RUN mv /app/.docker /build/
RUN chmod +x /build/.docker/*.sh && chmod +x /build/.docker/*/*.sh

#RUN composer self-update
#RUN cd /app && composer install --no-dev --prefer-dist --optimize-autoloader --no-interaction

# Give web server permissions to access app folder
RUN chown -R www-data:www-data /app

# run our uploads setup script
RUN bash /build/.docker/nginx/setup.sh

RUN crontab -u www-data /build/.docker/cron/drupalcron

# secure it by removing default keys.
RUN rm -rf /etc/service/sshd /etc/my_init.d/00_regen_ssh_host_keys.sh
```
- my `.docker` folder looks something like (image from a different project): http://screencast.com/t/m86OCyjykQYX .  This is essentially a per app version of https://github.com/that0n3guy/docker/tree/master/baseimage-nginx-phpfpm so I can modify the default nginx.conf and site-default for nginx... 
- use nginx from BOA probably
- use a host volume for `/app/sites/default/files` folder to the host server (someplace like /data/appdata/site1, Then backup the host `/data` folder to s3 with http://blog.phusion.nl/2013/11/11/duplicity-s3-easy-cheap-encrypted-automated-full-disk-backups-for-your-servers/ (a little outdated, but super easy to setup).  This to me is way simpler than using some drupal module to setup s3 buckets per drupal site.  You backup all dokku apps files to s3 and don't have to do it individually.  Easily movable to another server.

I'll probably have some time to do this over Christmas break.  I'll probably also create a docker hub image for drupal (IE, nginx already setup for drupal).

I'll create some more documentation later.

**Comment by mc0e** [comment]
I don't think you want to put the whole of drupal into your git repository.  Rather you want a builder that takes a drush make file and uses it to pull in the appropriate version of drupal and modules, apply any required patches to this contributed material.  Your git repository should only need to include the code you actually write yourself, and the drush makefile.

Given this drush make builder, a significant part of aegir's functionality is replaced.  It does provide a nice interface for cloning sites for testing of upgrades and so forth.  I wonder if Aegir's still the way to go for that, or some other approach.

**Comment by niccolox** [comment]
funnily enough I am just testing http://getvalkyrie.github.io/valkyrie/ which is Aegir 3

**Comment by that0n3guy** [comment]
@mc0e I don't do enough drupal dev to care to mess with make files. I used them in the past and found it a pain in my butt :).  It is good practice, don't get me wrong... but I'll just keep the whole drupal install in my repo.

**Comment by mc0e** [comment]
@that0n3guy The problem with that is that you need to keep changing stuff in your repository as security updates come out.

**Comment by that0n3guy** [comment]
Your right drush make is nice... I just moved my first site to dokku.  Uses drush make, nginx, caching, supports boost, upload progress... stuff like that.

##### About the docker image:
Here is the docker image I use for it: https://registry.hub.docker.com/u/that0n3guy/baseimage-nginx-phpfpm-drupal7/  

Its using https://github.com/perusio/drupal-with-nginx for nginx config.  If anyone has a simpler way to do this that has caching/boost support, submit a patch.

#### Example to use with dokku-alt
Here is the example: https://github.com/that0n3guy/dokku-alt-drupal-example

It shows you how to add cron, your own nginx vhost/config files, custom modules/theme/libraries, use drush make, etc... 

Make sure you look at the .docker/dokku.md file: https://github.com/that0n3guy/dokku-alt-drupal-example/blob/master/.docker/dokku.md.

It shows that you can use env vars for your mysql info

**Comment by ayufan** [comment]
Nice write-up. Maybe we can open add it to docs as example integrations?

On Tue, Feb 17, 2015 at 9:29 PM, that0n3guy <notifications@github.com>
wrote:

> Your right drush make is nice... I just moved my first site to dokku. Uses
> drush make, nginx, caching, supports boost, upload progress... stuff like
> that.
> About the docker image:
>
> Here is the docker image I use for it:
> https://registry.hub.docker.com/u/that0n3guy/baseimage-nginx-phpfpm-drupal7/
>
> Its using https://github.com/perusio/drupal-with-nginx for nginx config.
> If anyone has a simpler way to do this that has caching/boost support,
> submit a patch.
> Example to use with dokku-alt
>
> Here is the example:
> https://github.com/that0n3guy/dokku-alt-drupal-example
>
> It shows you how to add cron, your own nginx vhost/config files, custom
> modules/theme/libraries, use drush make, etc...
>
> Make sure you look at the .docker/dokku.md file:
> https://github.com/that0n3guy/dokku-alt-drupal-example/blob/master/.docker/dokku.md
> .
>
> It shows that you can use env vars for your mysql info
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dokku-alt/dokku-alt/issues/142#issuecomment-74746645>.
>



-- 
Kamil Trzciński

ayufan@ayufan.eu
www.ayufan.eu

**Comment by that0n3guy** [comment]
For sure.  I should do a writeup on how I do octobercms and x2engine (crm) as well.  Both nginx, php, and dockerfile...

**Comment by IbnSaeed** [comment]
Excellent addition. Does it work with using any docker images, such as Apache instead of nginx? 

I already have a few images setup for docker but i would like to test out dokku-alt using vagrant on my windows machine, would it be possible ?

**Comment by that0n3guy** [comment]
You could use whatever docker image you want.  Just change your dokku image to add the "app" wherever your http root is.

**Comment by IbnSaeed** [comment]
I tried to run dokku with vagrant on windows 8.1

But it gave out errors.

I am looking for other tools

**Comment by mc0e** [comment]
I've been playing with the buildpack approach.  What I have isn't fully worked, but if someone else is starting on this path, get in touch.

I've yet to decide whether the buildpack is the way to go.  

* Dockerfiles are nice in terms of limiting how much needs to get rebuilt each time you redeploy.   
* using composer to pull together all the required system packages, mysql extensions and libraries is nice, and for example switching from nginx to apache is a one line change in the Procfile.
* I have a problem with the permissions model adopted by the buildpack approach.  I haven't really gone into it enough yet, but if I can't separate the owner of the php files from the user from the user running the php process, then that's a show stopper for me.

I suspect the best way forward will be to use a Dockerfile to replace much of what the buildpack does, but still fire up composer to do a lot of the work.

I'm currently using drush make to build drupal, but noticing that Drupal is moving more towards composer.  I suspect it's the way of the future, but I'm not sure that it's a mature option now.  More to learn here.

**Comment by niccolox** [comment]
I just saw the official Drupal Docker https://github.com/docker-library/docs/tree/master/drupal

how does this change things?

**Comment by that0n3guy** [comment]
I don't think it changes anything.  It uses apache (I use nginx), it downloads the tar file (I try to always use drush make).

A mater of preference.

**Comment by gandhiano** [comment]
It might also be relevant to follow/engage in the [discussions around AegirNG (4.0)](https://aegir.hackpad.com/AegirNG-Proposal-GtXP9pgygHh).
